### PR TITLE
data disk encryption

### PR DESCRIPTION
**Pull request by pankajosh**
1. data disk encryption
2. Volume notification service driver code.
3. placing VNS driving logic in ADE.

### CODE REVIEW

**Comment by bamuir** [comment]
@pankajosh can you give us some background on VNS and what the overall process does??

File: VMEncryption/main/handle.py [hunk 1]
```diff
@@ -1858,6 +1906,11 @@ def enable_encryption_all_in_place(passphrase_file, encryption_marker, disk_util
                                                                                    status_prefix=status_prefix)
 
             if encryption_result_phase == CommonVariables.EncryptionPhaseDone:
```
Comment (canfikret) [inline VMEncryption/main/handle.py:1908]: let's consider scenarios where encryption phase is disrupted and volume is partially encrypted. In this case, saving the settings might be good idea as we start?

**Review by canfikret — CHANGES_REQUESTED** [review]

File: VMEncryption/main/handle.py [hunk 2]
```diff
@@ -867,7 +901,8 @@ def handle_encryption(public_settings, encryption_status, disk_util, bek_util, e
 
     if extension_parameter.config_file_exists() and extension_parameter.config_changed():
         logger.log("Config has changed, updating encryption settings")
-        hutil.exit_if_same_seq()
+        if not vns_call:
```
Comment (pankajosh) [inline VMEncryption/main/handle.py:904]: skip when seq check when call is triggered from vns_call.
remove not.

File: VMEncryption/main/handle.py [hunk 3]
```diff
@@ -888,7 +923,8 @@ def handle_encryption(public_settings, encryption_status, disk_util, bek_util, e
             logger.log('Encryption marker exists. Calling Enable')
             enable_encryption()
         else:
-            hutil.exit_if_same_seq()
+            if not vns_call:
```
Comment (pankajosh) [inline VMEncryption/main/handle.py:926]: skip when seq check when call is triggered from vns_call.
remove not.

File: VMEncryption/main/handle.py [hunk 4]
```diff
@@ -1073,7 +1109,8 @@ def enable_encryption():
 
 def perform_migration(encryption_config, crypt_mount_config_util):
     logger.log("Migrate operation found. Starting migration flow.")
-    hutil.exit_if_same_seq()
+    if not vns_call:
```
Comment (pankajosh) [inline VMEncryption/main/handle.py:1112]: skip when seq check when call is triggered from vns_call.
remove not.

File: VMEncryption/main/VolumeNotificationService.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+             self.workingDirectory = os.path.join(os.getcwd(),'..')
+        #normalize the path
+        self.workingDirectory = os.path.normpath(self.workingDirectory)
+    
+    def _service_file(self):
+        '''service file path'''
+        return os.path.join(self.workingDirectory,CommonVariables.vns_service_file)
+    
+    def _tmp_service_file(self):
+        '''get tem service file path'''
```
[the remark this answers is not in the document]
Comment (pankajosh) [reply VMEncryption/main/VolumeNotificationService.py:52]: done

File: VMEncryption/main/DiskUtil.py [hunk 6]
```diff
@@ -114,6 +114,51 @@ def create_cleartext_key(self, mapper_name):
...
+        self.logger.log(msg="import_token: device: {0} LUKS2 token field, update for wrapped passphrase")
+        protector = ""
+        with open(passphrase_file,"rb") as protector_file:
+            protector_data = protector_file.read()
+            import base64
+            protector = base64.standard_b64encode(protector_data).decode('utf_8')
+        KekVaultResourceId=public_settings.get(CommonVariables.KekVaultResourceIdKey)
+        KeyEncryptionKeyUrl=public_settings.get(CommonVariables.KeyEncryptionKeyURLKey)
+        data={
+            "type":"Azure_Disk_Encryption",
```
[the remark this answers is not in the document]
Comment (pankajosh) [reply VMEncryption/main/DiskUtil.py:136]: agree done

File: VMEncryption/main/VolumeNotificationService.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+                os.remove(runningservicefilepath)
+            shutil.copy(self._tmp_service_file(),runningservicefilepath)
+            cmd = 'systemctl daemon-reload'
+            return_code = self.command_executor.Execute(cmd)
+            if return_code == 0:
+                #enable the service 
+                return_code = self.enable()
+        return return_code
+
+    def deRegister(self):
```
[the remark this answers is not in the document]
Comment (pankajosh) [reply VMEncryption/main/VolumeNotificationService.py:117]: done

File: VMEncryption/main/VolumeNotificationService.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    def status(self):
+        '''
+        ActiveState of service is active and inactive 
+        '''
+        cmd = 'systemctl show -p ActiveState --value '+ CommonVariables.vns_service_file
+        proc_comm=ProcessCommunicator()
+        self.command_executor.Execute(cmd,communicator=proc_comm)
+        status=proc_comm.stdout
+        return status.strip()
```
[the remark this answers is not in the document]
Comment (pankajosh) [reply VMEncryption/main/VolumeNotificationService.py:156]: done

File: VMEncryption/main/handle.py [hunk 9]
```diff
@@ -67,11 +68,17 @@ def disable():
     # to restore these configs in their update step rather than their install step once all
     # released versions of the extension are at this version or above
     hutil.archive_old_configs()
+    if security_Type == CommonVariables.ConfidentialVM:
+        vns_service = VolumeNotificationService(logger=logger)
+        vns_service.stop()
```
[the remark this answers is not in the document]
Comment (pankajosh) [reply VMEncryption/main/handle.py:73]: done, mask on disable and unmask on enable.

File: VMEncryption/main/handle.py (hunk 2 above)
Comment (pankajosh) [reply VMEncryption/main/handle.py:904]: done

File: VMEncryption/main/handle.py (hunk 3 above)
Comment (pankajosh) [reply VMEncryption/main/handle.py:926]: done

File: VMEncryption/main/handle.py (hunk 4 above)
Comment (pankajosh) [reply VMEncryption/main/handle.py:1112]: done

File: VMEncryption/main/patch/UbuntuPatching.py [hunk 10]
```diff
@@ -46,6 +46,8 @@ def __init__(self, logger, distro_info):
         self.resize2fs_path = '/sbin/resize2fs'
         self.umount_path = '/bin/umount'
         self.touch_path = '/usr/bin/touch'
+        self.min_version_online_encryption='22.04'
+        self.support_online_encryption=self.validate_online_encryption_support()
```
Comment (pankajosh) [inline VMEncryption/main/patch/UbuntuPatching.py:50]: [vishal]: We don't support online encryption on Ubuntu as the OS disk does not have a separate boot partition.
So based on the above comment support_online_encryption must not be true, whereas 
if CVM and self.validate_online_encryption_support()
     do online data disk encryption.

**Review by canfikret — APPROVED** [review]

**Pull request merged**